File: usvao/VAO/software/datascope/projects/query/tags/080414.r242/trunk/src/java/net/ivoa/query/BaseQuery.java
package net.ivoa.query;

import net.nvo.Header;
import net.nvo.Footer;

import net.ivoa.util.CGI;
import net.ivoa.util.Settings;

public class BaseQuery {
	public static void main(String[] args) {
		CGI params = new CGI();
		if (Querier.validQuery(params)) {
			Querier.run(params);
		} else {
			run(params);
		}
	}

	static void run(CGI cgi) {
		BaseQuery.run(cgi,"");
	}

	static void run(CGI cgi, String errorMessage) {
		for (String key: cgi.keys() ) {
			if (!(key.equals("sources") || key.equals("resources")) || cgi.value(key).length() > 1) {
				Settings.put(key, cgi.value(key));
			}
		}

		Header h = new Header();
		h.setBannerTitle("NVO Portal Services: Simple Data Query");
		h.setTitle("Simple Query");

		//	h.addCSSFile(Settings.get("DocBase")+"/css/styles.css");
		h.addCSSFile(Settings.get("DocBase")+"/css/style.css");
		h.addJavaScript(Settings.get("docbase")+"/js/jquery.pack.js");
		h.addJavaScript(Settings.get("docbase")+"/js/jquery.listen-min.js");
		h.addJavaScript(Settings.get("docbase")+"/js/jquery.simplemodal.js");
		h.addJavaScript(Settings.get("docbase")+"/js/validate.js");
		h.addJavaScript(Settings.get("docbase")+"/js/basequery.js");

		h.addToken("http://us-vo.org", "NVO&nbsp;home&nbsp;page");
		h.addToken("/cgi-bin/vo/squery/query.sh", "New Query", "Begin new query");
		h.addToken("http://nvo.stsci.edu/vor10/index.aspx", "Registry", "Search for resources using keywords");
		h.addToken("http://irsa.ipac.caltech.edu/applications/VOInventory", "Inventory", "Search for resources that have coverage for a list of positions");
		h.addToken("http://heasarc.gsfc.nasa.gov/cgi-bin/vo/datascope/init.pl", "DataScope", "Search for all data on a given position/object");
		h.addToken("http://nesssi.cacr.caltech.edu/cgi-bin/vim.cgi", "VIM",  "Search and combine data from selected resources and targets");
		h.addToken(Settings.get("DocBase")+"/scripting.html", "Scripting", "How to run retrieval scripts on your machine");
		h.addToken(Settings.get("DocBase")+"/helpInc.html", "Help", "Help for the SimpleQuery service");
		h.addToken("http://us-vo.org/feedback/index.cfm", "NVO&nbsp;Feedback");

		h.printHTTP(System.out);
		h.printHTMLHeader(System.out);
		System.out.println("<div id='content'>");
		h.printBanner(System.out);
		System.out.println("<p><h2 id='querycaption'>");
		if (Settings.has("ivoid")  && Settings.has("shortname")) {
			System.out.println("Query resource: "+Settings.get("shortname"));
		} else {
			System.out.println("Query a single NVO resource");
		}
		System.out.println("</h2></p>");
		printForm(errorMessage);
		System.out.println("</div><!-- id='content' -->");
//		System.out.println("<div id='footer'>");
		new Footer().print(System.out);
//		System.out.println("</div><!-- id='footer' -->");// this actually comes after the closing html tag
	}

	static void printForm() {
		printForm("");
	}

	static void printForm(String errorMessage) {
		String sname = Settings.get("shortname");
//		String ivoid    = null;	//Settings.get("ivoid");
//		String hidden= "";

//		// added to deal with empty parameter (which is NOT null)
//		if ( Settings.has("ivoid") && Settings.get("ivoid").length() >0 ) {
//			ivoid = Settings.get("ivoid");
//		}

//		boolean needPos = true;

		System.out.printf(
			"<form id='basequery' action='%s/query.sh' onsubmit='return validate(this);' method='POST' enctype='multipart/form-data' >\n", Settings.get("CGIBase"));

		System.out.println("<table id='formtable' width='100%' style='border:1px solid silver;'>");

		// for javascript error message
		System.out.println("<tr><td id='errorHolder'>");
		//	Java error message
		if ( errorMessage.length() > 0 ) {
			System.out.println("<p class='errorMessage' >"+errorMessage+"</p>");
		}
		System.out.println("</td></tr>");

		System.out.println("<tr><td><table id='querytable' width='100%' style='border:1px solid silver;'>");

		System.out.println(
			"<tr><td colspan='2' align='center' id='querytitle' >Query ...</td></tr>\n"+
			"<tr><td colspan='2'><table id='positiontable' width='100%' style='border:1px solid silver;'>\n"+
			"<tr id='trPOSITION'>\n"+
			"<td align='center'><span id='positionmodal'>Position:</span><input class='src' size='30' id='POSITION' name='POSITION' value='"+Settings.get("POSITION", "")+"' /></td>\n" +
			"</tr>\n"+
			"<tr><td align='center'>- OR -</td></tr>\n"+
			"<tr id='trallsky'><td align='center'><span id='allskymodal'>All-sky</span> <input class='src' id='allsky' name='allsky' type='checkbox' value='checked' /></td></tr>\n"+
			"</table><!-- id='positiontable' --></td></tr>");

		System.out.println(
			"<tr id='trRADIUS'><td colspan='2' align='center'>Radius:<input id='RADIUS' size='5' name='RADIUS' value='"+Settings.get("Radius", "")+"' />\n"+
			"<select id='units' name='units'>");
		String units = Settings.get("units","degree");
		String[] options = {"degree","arcmin","arcsec"};
		for (String option: options) {
			String selected = "";
			if ( units.equals(option) ) selected = " selected='true'";
			System.out.println("<option"+selected+">"+option+"</option>");
		}
		System.out.println( "</select>\n</td></tr>");
/*
		if (ivoid != null) {
			hidden = " style='display: none'";
			if (sname != null) {
				System.out.println("<input type='hidden' name='ShortName' value='"+sname+"' />");
			}
		}
*/
		System.out.println(
			"<tr id='trIVOID'>\n"+
			"<td align='center'><span id='ivoidmodal'>IVO Identifier:</span><input size='30' id='IVOID' name='IVOID' value='"+Settings.get("ivoid", "")+"' /></td>\n"+
			"<td align='center'><span id='alldatamodal'>All data?</span> <input id='Verbosity' type='checkbox' name='Verbosity' /></td>\n"+
			"</tr>");

		System.out.println( "</table><!-- id='querytable' --></td></tr>");

		System.out.println(
			"<tr><td><table id='viewtable' width='100%' style='border:1px solid silver;'>\n"+
			"<tr><td colspan='2' align='center'>or View ...</td></tr>\n"+
			"<tr id='trviewLocal'><td align='right'>Local VO table</td><td><input size='30' class='src' id='viewLocal' name='viewLocal' type='file' /></td></tr>\n"+
			"<tr><td colspan='2' align='center'>- OR -</td></tr>\n"+
			"<tr id='trviewURL'><td align='right'>VO table URL</td><td><input size='30' class='src' id='viewURL' name='viewURL' /></td></tr>\n"+
			"</table><!-- id='viewtable' --></td></tr>");


		System.out.println(
			"<tr><td align='center'><input type='submit' /><input id='reset' type='reset' /></td></tr>\n"+
			"</table><!-- id='formtable' -->");

		if (Settings.has("requestID")) {
			System.out.println("<input type='hidden' name='requestID' value='"+Settings.get("requestID")+"' />");
		}


		System.out.println("<input class='src' id='sourcesURL' type='hidden' name='sourcesURL' value='"+Settings.get("sourcesURL","")+"' />");

		String resourcesURL = Settings.get("resourcesURL");
		if (valid(resourcesURL)) {
			System.out.println("<input type='hidden' name='resourcesURL' value='"+resourcesURL+"' />");
//	Jake - I changed this because of the mismatch (assuming it wasn't done on purpose)
//			System.out.println("<input type=hidden name=sourcesURL value='"+resourcesURL+"'>");
		}

		String resources = Settings.get("resources");
		if (valid(resources) ) {
			System.out.println("<input type='hidden' name='resources' value='"+encodeXML(resources)+"' />");
		}

		System.out.println("<input id='sources' class='src' type='hidden' name='sources' value='"+encodeXML(Settings.get("sources",""))+"' />");

		System.out.println("<input type='hidden' name='limit' value='5' />");
		System.out.println("</form>");
	}

	private static boolean valid(String input) {
		return input != null && input.length() > 0;
	}

	private static String encodeXML(String input) {
		// Replace single quotes and newlines so
		// that we can put this in hidden variable.
		// The encoding should be undone when we
		// interpret the value in JavaScript.
		input = input.replaceAll("'", "\\'");
		input = input.replaceAll("\n", "\\n");
		return input;
	}
}

File: usvao/VAO/software/datascope/projects/query/trunk/src/javascript/basequery.js
// Functions related to be the basic functionality of querying
// in the SimpleQuery tool.

var candidateIVOID;
var candidateIVOIDIndex;
var myModal;

// Define functionality associated with actions on the form.
$(function() {

    // Get the metadata for a resource.	
    $.listen('click','#getmetadata', function(){
	$.get("@URL_PATH@metaquery.pl", {IVOID: $('#IVOID').val() }, function(data){
	     myModal = $(data).modal({close:true});
        });
    });
	
    // If the user leaves the IVOID identifier field, then
    // update the caption for the field.
	
    $.listen('blur','#IVOID', updateCaption );

    // If the user leaves the inputs source function then ???
    $('input.src').blur( function() {
	if ( ( $(this).val().length <= 0 ) || ( ( ! $(this).attr('checked') ) && ( $(this).val() == "checked" ) ) ) {
	    $('input.src').attr('disabled',false);
	} else {
	    $('input.src').attr('disabled',true);
	    $(this).attr('disabled',false);
	}
    });

    // If the user resets the inputs then clean up.
    $('input#reset').click( function() {
        $('#errorHolder').empty();
        $('form#basequery tr').removeClass('errorField');
	$('input.src').attr('disabled',false);
	$('input.src').val(''); // clear out hidden src parameters (sources, sourcesURL) as well
      	$('input.src').fadeIn();
	$('#IVOID').val('');
	$('#viewtable').fadeIn();
	$('#positiontable').fadeIn();
        updateCaption();
    });

    $('#positionmodal').click(function () {
	$.get("@URL_PATH@position.html", function(data){
	    $(data).modal({ });
        });
    });
    
    $('#radiusmodal').click(function () {
        $.get("@URL_PATH@radius.html", function(data){
	     $(data).modal({ });
	});
    });
    
    $('#alldatamodal').click(function () {
	$.get("@URL_PATH@alldata.html", function(data){
	    $(data).modal({ });
        });
    });
        
    $('#allskymodal').click(function () {
        $.get("@URL_PATH@allsky.html", function(data){
	    $(data).modal({ });
        });
    });
    
    $('#ivoidmodal').click(function () {
	$.get("@URL_PATH@ivoid.html", function(data){
	    $(data).modal({ });
	});
    });
    
    $('#querytoggle').click(function () {
        $('#innerquerydiv').slideDown();
        $('#innerviewdiv').slideUp();
    });
    
    $('#viewtoggle').click(function () {
        $('#innerviewdiv').slideDown();
        $('#innerquerydiv').slideUp();
    });
    
    updateCaption();
    $('#innerviewdiv').slideUp();
    
    if ( $('#POSITION').val() || $('#sources').val() || $('#sourcesURL').val() ) {
	$('#viewtable').fadeOut();
//      $('input.src').fadeOut();
//	$('#querytitle').append('<br/>Using position(s) from given POSITION/sources/sourcesURL');
    }
    
    if ( $('#IVOID').val() ) {
	$('#viewtable').fadeOut();
    }
});


function doUpdateCaption(index) {
    var ivo = candidateIVOID[index].split(/\s+/);
    $('#IVOID').val(ivo.shift()); // insert complete IVOID as the search will match partial
    $('#querycaption').html("Query resource:&nbsp;"+ivo.join(' ')+"&nbsp;(<a class='small' href='javascript:void(0)' id='getmetadata'>Columns</a>)");
    candidateIVOIDIndex = index;
}
    
function updateCaption () {
    if ( $('#IVOID').val() ) {
        $.get('@URL_PATH@getIVOnames.pl', { IVOID: $('#IVOID').val() },
          function(data){
	      if (data) {
		  candidateIVOID = data.split("\n");
		  if (candidateIVOID[candidateIVOID.length-1].length == 0) {
		      candidateIVOID.pop();
		  }
		  if (candidateIVOID.length > 0) {
		      doUpdateCaption(0);
		  }
		  if (candidateIVOID.length > 1) {
		      showCandidates();
		  }
              }
	  }
	);
    } else {
        candidateIVOID      = null;
	candidataIVOIDIndex = null;
    }
}

function showCandidates() {

    var str = "<td><td><a href='javascript:void(0)' id='ivoidmodal'>Multiple IVO Identifiers</a><br> Please select a table:</td>"+
                   "<td><select name='IVOID' id='IVOID' onchange='doUpdateCaption(this.selectedIndex); return true;'> ";
    for (var i=0; i<candidateIVOID.length; i += 1) {
        var ivo = candidateIVOID[i].split(/\s+/);
	
	var selected = "";
	if (i == 0) {
	    selected = " selected ";
	}
	var id = ivo[0];
	id = id.replace(/^.*VizieR\//g,"");
        str += "<option value='" + ivo[0] + "'" + selected + "> " + id+"\n";
    } 
    str += "</select></td>"+
           "<td align='center'><a href='javascript:void(0)' id='alldatamodal'>All columns?</a> <input id='Verbosity' type='checkbox' name='Verbosity' /></td>";
    document.getElementById("trIVOID").innerHTML = str;
}
